src/Strategy/ManifestStrategy.php

```diff
@@ -0,0 +1,178 @@
+        $versionInfo = $this->getAvailableVersions();
+        if (!isset($versionInfo[$version])) {
+            throw new \RuntimeException(sprintf('Failed to find manifest item for version %s', $version));
+        }
+
+        $fileContents = file_get_contents($versionInfo[$version]['url']);
+        if ($fileContents === false) {
+            throw new HttpRequestException(sprintf('Failed to download file from URL: %s', $versionInfo[$version]['url']));
+        }
+
+        $tmpFilename = $updater->getTempPharFile();
+        if (file_put_contents($tmpFilename, $fileContents) === false) {
+            throw new \RuntimeException(sprintf('Failed to write file: %s', $tmpFilename));
+        }
+
+        $tmpSha = sha1_file($tmpFilename);
+        if ($tmpSha !== $versionInfo[$version]['sha1']) {
+            unlink($tmpFilename);
+            throw new \RuntimeException(
+                sprintf(
+                    'SHA-1 verification failed: expected %s, actual %s',
+                    $versionInfo[$version]['sha1'],
+                    $tmpSha
+                )
+            );
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCurrentRemoteVersion(Updater $updater)
+    {
+        $versions = array_keys($this->getAvailableVersions());
+        if (!$this->allowMajor) {
+            $versions = $this->filterByLocalMajorVersion($versions);
+        }
+
+        $versionParser = new VersionParser($versions);
+
+        $mostRecent = $versionParser->getMostRecentStable();
+
+        // Look for unstable updates if explicitly allowed, or if the local
+        // version is already unstable and there is no new stable version.
+        if ($this->allowUnstable || ($versionParser->isUnstable($this->localVersion) && version_compare($mostRecent, $this->localVersion, '<'))) {
+            $mostRecent = $versionParser->getMostRecentAll();
+        }
+
+        return version_compare($mostRecent, $this->localVersion, '>') ? $mostRecent : false;
+    }
+
+    /**
+     * Get available versions to update to.
+     *
+     * @return array
+     *   An array keyed by the version name, whose elements are arrays
+     *   containing version information ('name', 'sha1', and 'url').
+     */
+    private function getAvailableVersions()
+    {
+        if (!isset($this->availableVersions)) {
+            $this->availableVersions = array();
+            foreach ($this->getManifest() as $key => $item) {
+                if ($missing = array_diff(self::$requiredKeys, array_keys($item))) {
+                    throw new \RuntimeException(sprintf('Manifest item %s missing required key(s): %s', $key, implode(',', $missing)));
+                }
+                $this->availableVersions[$item['version']] = $item;
+            }
+        }
+
+        return $this->availableVersions;
+    }
+
+    /**
+     * Download and decode the JSON manifest file.
+     *
+     * @return array
+     */
+    private function getManifest()
```

Comment:
`retrieveManifest()`? I dislike `get*` here because although natural it's too mixed with the getter/setter idea which are 90% just about returning a property without any logic. `retrieveManifest()` kinda conveys that it's not just about returning a property value.